### ISSUE TITLE
Crash while creating sea ice initial conditions

### ISSUE THREAD

**Issue by DanJonesOcean** [so-wise/so-wise-gyre#8]
I've been able to fix the issues with the theta and salt initial conditions, but I am still having an issue with the sea ice initial conditions. Here is the error message from the `sose_ics` function:

```
...interpolating to model grid
Writing initial_conditions/SALT_BSOSE.ini
...adding to initial_conditions/sose_ics.nc
Processing SIarea
...reading /data/oceans_input/raw_input_data/SOSE_monthly_climatology/SIarea_climatology.data
/data/oceans_input/raw_input_data/SOSE_monthly_climatology/SIarea_climatology.meta
...extrapolating into missing regions
......259137 points to fill
Error (discard_and_fill): some missing values cannot be filled
An exception has occurred, use %tb to see the full traceback.
```

Here is the full traceback:

```
In [34]: %tb
---------------------------------------------------------------------------
SystemExit                                Traceback (most recent call last)
<ipython-input-33-20c73ff3dcbb> in <module>()
      1 
----> 2 sose_ics(grid_path='grid/', sose_dir='/data/oceans_input/raw_input_data/SOSE_monthly_climatology/', nc_out='initial_conditions/sose_ics.nc', output_dir='initial_conditions/', constant_t=-1.9, constant_s=34.4, split=180)

/data/expose/so-wise/so-wise-setup/mitgcm_python/ics_obcs.py in sose_ics(grid_path, sose_dir, output_dir, bsose, nc_out, constant_t, constant_s, split, prec)
    191             # Fill any missing regions with zero sea ice, as we won't be extrapolating them later
    192             sose_data = sose_grid.read_field(in_file, 'xyt', var_name=fields[n], fill_value=0)[0,:]
--> 193         process_ini_field(sose_data, sose_mask, fill, sose_grid, model_grid, dim[n], fields[n], out_file, model_cavity=model_cavity, cavity_value=constant_value[n], nc_out=nc_out, ncfile=ncfile, prec=prec)
    194 
    195     if nc_out is not None:

/data/expose/so-wise/so-wise-setup/mitgcm_python/ics_obcs.py in process_ini_field(source_data, source_mask, fill, source_grid, model_grid, dim, field_name, out_file, missing_cavities, model_cavity, cavity_value, regular, nc_out, ncfile, prec, missing_val)
     90     else:
     91         # Just the surface layer
---> 92         source_data = discard_and_fill(source_data, source_mask[0,:], fill[0,:], use_3d=False, missing_val=missing_val)
     93 
     94     print '...interpolating to model grid'

/data/expose/so-wise/so-wise-setup/mitgcm_python/interpolation.py in discard_and_fill(data, discard, fill, missing_val, use_1d, use_3d, preference, log)
    344             if num_missing == num_missing_old:
    345                 print 'Error (discard_and_fill): some missing values cannot be filled'
--> 346                 sys.exit()
    347     return data
    348 

SystemExit: 
```

Let's see what we can do.

**Comment by DanJonesOcean** [comment]
As with the temperature and salinity, it may be the case that there are small ice shelf points that could be removed. Let's work on that and see how it goes.

**Comment by DanJonesOcean** [comment]
Removing ice shelf points did not get rid of this error. I'm using "embed" statements to debug the code. That is, I'm inserting lines of the form `from IPython import embed; embed()` into the code where I need to embed. I've tracked it down to this function call so far:

```
  92         from IPython import embed; embed()
  93         source_data = discard_and_fill(source_data, source_mask[0,:], fill[0,:], use_3d=False, missing_val=missing_val)
```
The `source_data` variable looks okay:

![image](https://user-images.githubusercontent.com/11757453/112879932-bd109300-90c1-11eb-81a1-9bd945a69b9f.png)

I need to learn more about the `source_mask`. The top layer `source_mask[0,:]` The top layer appears to be all zeros, but the layer below it has some non-zero values. Next step: figure out what the source mask does and where it comes from.

**Comment by DanJonesOcean** [comment]
I'm finally returning to this issue after a long absence. I'm still having trouble with the "discard and fill" function. 

@knaughten, could you please say a little bit about what the "source_mask" field does? Where does it come from?

**Comment by DanJonesOcean** [comment]
For reference, here are some inputs into the `discard_and_fill` function:

```
In [13]: source_data.max()
Out[13]: 0.99387389421463013

In [14]: source_data
Out[14]: 
array([[ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       ..., 
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.]])

In [15]: source_mask[0,:]
Out[15]: 
array([[ True,  True,  True, ...,  True,  True,  True],
       [ True,  True,  True, ...,  True,  True,  True],
       [ True,  True,  True, ...,  True,  True,  True],
       ..., 
       [ True,  True,  True, ...,  True,  True,  True],
       [ True,  True,  True, ...,  True,  True,  True],
       [ True,  True,  True, ...,  True,  True,  True]], dtype=bool)

In [16]: fill[0,:]
Out[16]: 
array([[ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       ..., 
       [ 0.,  0.,  0., ...,  1.,  1.,  1.],
       [ 0.,  0.,  0., ...,  1.,  1.,  1.],
       [ 0.,  0.,  0., ...,  1.,  1.,  1.]])

In [17]: missing_val
Out[17]: -9999
```

Once inside the `discard_and_fill` function, we run into the following block of code:
```
  if num_missing == num_missing_old:
      print 'Error (discard_and_fill): some missing values cannot be filled'
      from IPython import embed; embed()
      sys.exit()
```
The "embed" statement is for debugging purposes only. Here is where we would like some file I/O to output the problem locations. Kaitlin suggests that we use the class NCfile in file_io.py. Here's a typical usage of that class: 
```
      ncfile = NCfile(nc_out, grid, 'xyz')
      ncfile.add_variable('THETA', temp, 'xyz')
      ncfile.add_variable('SALT', salt, 'xyz')
      ncfile.close()
```
We can use similar syntax for our issues. We can kind of hard code the filename for the error file, but not grid. We'll need the `discard_and_fill` function to have access to `grid`.

**Comment by DanJonesOcean** [comment]
Okay, some debugging has brought me here:
- When `discard_and_fill` is first called, the input fields `discard`, and `fill` already have some issues.  
- `source_data`, which is passed to the function as `data`, looks okay outside the function and within.
- `source_mask[0,:]`, which is passed to the function as `discard`, is true everywhere, indicating that every point should be discarded. Where did this come from?
- `fill[0,:]`, which is passed to the function as `fill`, looks very bizarre (see below).

![fill](https://user-images.githubusercontent.com/11757453/119514496-f9343d00-bd6c-11eb-80a4-f9c472709b88.png)

So what's going on? Let's trace this further back in the code to see what's up. In the meantime, we can include this debugging/output statement/dump.

**Comment by DanJonesOcean** [comment]
These statements are so good for debugging:

`from IPython import embed; embed()`

**Comment by DanJonesOcean** [comment]
Anyway, these fields look weird even when called by `discard_and_fill`. This function is called by `process_ini_field`, which is called by `sose_ics`. More debugging!

```
In [5]: np.shape(sose_mask)
Out[5]: (44, 327, 1052)
```

The top level (`k=0`) of the mask is zero everywhere. The next level down (`k=1`) has the continents. Where did these dimensions come from? 

Here's the min/max:
```
In [30]: sose_mask[0,:,:].min()
Out[30]: True

In [31]: sose_mask[1,:,:].min()
Out[31]: False
```
This comes from the `sose_grid` object:
```
In [37]: sose_grid.hfac[0,:,:].max()
Out[37]: 0.0
```
which comes from:
```
    sose_grid = SOSEGrid(sose_grid_path, model_grid=model_grid, split=split)
```
Check out the input parameters of this function. There's some guidance in the file itself:
```
sose_ics (grid_path, sose_dir, output_dir, bsose=False, nc_out=None, constant_t=-1.9, constant_s=34.4, split=180, prec=64):
```
The split value of 180 should be fine, because our domain includes 0° longitude. And I think `bsose=False` is correct too.

Okay, so the `fill` field also looks odd. It comes from here:
```
fill, model_cavity = get_fill_mask(sose_grid, model_grid)
```
I'm guessing this means that there is some inherent problem in the model grid. This makes me think that I should start with something that I know works and extend it northwards. I should maybe start over and try KN's grid for the 1/4° case...

**Comment by DanJonesOcean** [comment]
Here's the relevant helper function:
```
# Helper function for initial conditions: figure out which points on the source grid will be needed for interpolation. Does not include ice shelf cavities, unless missing_cavities=False.
def get_fill_mask (source_grid, model_grid, missing_cavities=True):

    from interpolation import interp_reg

    # Find open cells according to the model, interpolated to source grid
    model_open = np.ceil(interp_reg(model_grid, source_grid, np.ceil(model_grid.hfac), fill_value=0))
    if missing_cavities:
        # Find ice shelf cavity points according to model, interpolated to source grid
        model_cavity = np.ceil(interp_reg(model_grid, source_grid, xy_to_xyz(model_grid.ice_mask, model_grid), fill_value=0)).astype(bool)
        # Select open, non-cavity cells
        fill = model_open*np.invert(model_cavity)
    else:
        fill = model_open
    # Extend into the mask a few times to make sure there are no artifacts near the coast
    fill = extend_into_mask(fill, missing_val=0, use_3d=True, num_iters=3)
    if missing_cavities:
        return fill, model_cavity
    else:
        return fill
```
I've put a debugging statement in just before the final `if` statement. The field `model_open` looks odd. It's zero on the top layer and then looks odd one layer down. This appears to be the line where the errors are introduced:
```
model_open = np.ceil(interp_reg(model_grid, source_grid, np.ceil(model_grid.hfac), fill_value=0))
```
Note that the fill value is zero. Also note that there is no `dim` parameter, so it will be 3 by default. Is that appropriate for the sea ice fields? Also, why are model grid and source grid between the definition below and the call above? 
```
# Interpolate a field on a regular MITgcm grid, to another regular MITgcm grid. Anything outside the bounds of the source grid will be filled with fill_value.
# source_grid and target_grid can be either Grid or SOSEGrid objects.
# Set dim=3 for 3D fields (xyz), dim=2 for 2D fields (xy).
def interp_reg (source_grid, target_grid, source_data, dim=3, gtype='t', fill_value=-9999):
```
Finally, this is called:
```
# Interpolate from a regular lat-lon grid to another regular lat-lon grid.
# source_lon and source_lat should be 1D arrays; target_lon and target_lat can be either 1D or 2D.
# Fill anything outside the bounds of the source grid with fill_value, but assume there are no missing values within the bounds of the source grid.
def interp_reg_xy (source_lon, source_lat, source_data, target_lon, target_lat, fill_value=-9999):

    from scipy.interpolate import RegularGridInterpolator

    # Build an interpolant
    interpolant = RegularGridInterpolator((source_lat, source_lon), source_data, bounds_error=False, fill_value=fill_value)
    if len(target_lon.shape) == 1:
        # Make target lat/lon arrays 2D
        target_lon, target_lat = np.meshgrid(target_lon, target_lat)
    # Interpolate
    data_interp = interpolant((target_lat, target_lon))
    return data_interp
```
More debugging statements to follow.

**Comment by DanJonesOcean** [comment]
Here's the traceback of where it crashes:

```
In [4]: %tb
---------------------------------------------------------------------------
SystemExit                                Traceback (most recent call last)
<ipython-input-3-40ca1008169b> in <module>()
      1 from mitgcm_python.ics_obcs import *
----> 2 sose_ics(grid_path='../grid/', sose_dir='/data/oceans_input/raw_input_data/SOSE_monthly_climatology/', nc_out='initial_conditions/sose_ics.nc', output_dir='initial_conditions/', constant_t=-1.9, constant_s=34.4, split=180)

/data/expose/so-wise/so-wise-gyre/setup/mitgcm_python/ics_obcs.py in sose_ics(grid_path, sose_dir, output_dir, bsose, nc_out, constant_t, constant_s, split, prec)
    194             # Fill any missing regions with zero sea ice, as we won't be extrapolating them later
    195             sose_data = sose_grid.read_field(in_file, 'xyt', var_name=fields[n], fill_value=0)[0,:]
--> 196         process_ini_field(sose_data, sose_mask, fill, sose_grid, model_grid, dim[n], fields[n], out_file, model_cavity=model_cavity, cavity_value=constant_value[n], nc_out=nc_out, ncfile=ncfile, prec=prec)
    197 
    198     if nc_out is not None:

/data/expose/so-wise/so-wise-gyre/setup/mitgcm_python/ics_obcs.py in process_ini_field(source_data, source_mask, fill, source_grid, model_grid, dim, field_name, out_file, missing_cavities, model_cavity, cavity_value, regular, nc_out, ncfile, prec, missing_val)
     90     else:
     91         # Just the surface layer
---> 92         source_data = discard_and_fill(source_data, source_mask[0,:], fill[0,:], use_3d=False, missing_val=missing_val)
     93 
     94     print '...interpolating to model grid'

/data/expose/so-wise/so-wise-gyre/setup/mitgcm_python/interpolation.py in discard_and_fill(data, discard, fill, missing_val, use_1d, use_3d, preference, log)
    358                 fio.write_netcdf_very_basic(discard, 'discard', 'error_dump_discard.nc', use_3d=use_3d)
    359                 fio.write_netcdf_very_basic(fill,    'fill',    'error_dump_fill.nc', use_3d=use_3d)
--> 360                 sys.exit()
    361     return data
    362 
```

**Comment by DanJonesOcean** [comment]
The inputs to `process_ini_field`:

- source_data: looks okay
- sose_mask: True everywhere
- fill: top layer looks odd
- model_cavity: top layer is False everywhere
- constant_value[n]: zero

So we're back where we started. What generates the 'sose_mask' and 'fill'? And why is it working for some fields and not others? Again, fill comes from
```
fill, model_cavity = get_fill_mask(sose_grid, model_grid)
```
which calls:
```
model_open = np.ceil(interp_reg(model_grid, source_grid, np.ceil(model_grid.hfac), fill_value=0))
```
which calls:
```
interp_reg_xyz(source_lon, source_lat, source_grid.z, source_data, target_lon, target_lat, target_grid.z, fill_value=fill_value)
```
Something odd is happening in here:
```
# Like interp_reg_xy, but for lat-lon-depth grids.
def interp_reg_xyz (source_lon, source_lat, source_z, source_data, target_lon, target_lat, target_z, fill_value=-9999):

    from scipy.interpolate import RegularGridInterpolator

    # Build an interpolant
    # Make depth positive so it's strictly increasing
    interpolant = RegularGridInterpolator((-source_z, source_lat, source_lon), source_data, bounds_error=False, fill_value=fill_value)
    # Make target axes 3D    
    if len(target_lon.shape) == 1:
        target_lon, target_lat = np.meshgrid(target_lon, target_lat)
    dimensions = [target_lon.shape[1], target_lon.shape[0], target_z.size]
    target_lon = xy_to_xyz(target_lon, dimensions)
    target_lat = xy_to_xyz(target_lat, dimensions)
    target_z = z_to_xyz(target_z, dimensions)
    # Interpolate
    data_interp = interpolant((-target_z, target_lat, target_lon))
    return data_interp
```
That's where we should be adding debugging statements, in that function in the `interpolation.py` function. I'm guessing that it's my model grid causing problems. Perhaps a shift to a different grid would help.

**Comment by DanJonesOcean** [comment]
**I've tracked down the error.** The function `interp_reg_xyz` attempts to interpolate the SOSE initial conditions onto the SO-WISE grid. However, the very top layer of the SO-WISE grid is "outside the bounds of the source grid". The problem occurs in the vertical interpolation. 

The top vertical cell of the SOSE grid has a vertical coordinate of -2.5. The top vertical cell of the SO-WISE grid has a vertical coordinate of 0.0. So we need to redesign the vertical grid! Just slightly.

**Comment by DanJonesOcean** [comment]
Checking again. The SO-WISE grid: the Z looks okay, but for some reason Zl (vertical coordinate of upper cell interface) is enormous (1e36). Next, let's figure out what it is about the SO-WISE and SOSE vertical grids that seem to confuse the interpolation process.

**Comment by knaughten** [comment]
Hi Dan, thanks for your patience while I finished up a few high-priority jobs. It looks to me like the SO-WISE grid is the culprit: the uppermost Zl should be 0 (the sea surface), not the uppermost Z. Let me know if you are still stuck and I'll see what I can do.

**Comment by DanJonesOcean** [comment]
Hi @knaughten. Now that half-term break is over, I'm investigating the vertical grid. When I examine the output files from MITgcm, everything looks okay. That is, 
```
RF(1) = 0
RF(2) = -5

RC(1) = -2.5
```
Those vertical coordinates are consistent with a grid with a top edge at the sea surface (0.0) and has a top cell thickness of 5m. The central coordinate is 2.5m below the surface, which is also correct. So why does `mitgcm_python` seem to have the incorrect grid? I'll see what I can find.

**Comment by DanJonesOcean** [comment]
By the way, the `mitgcm_python` function `check_final_grid` returns "Everything looks good!". The SO-WISE grid at least passes those checks.

**Comment by DanJonesOcean** [comment]
Just before the function call to `process_ini_field` that will create the sea ice initial conditions, we have the following:
```
In [6]: sose_grid.z[0:4]
Out[6]: array([  0. ,  -5. , -15.5, -27. ])

In [7]: sose_grid.z_edges[0:4]
Out[7]: array([  0.,   0., -10., -21.])
```
and the model grid:
```
In [9]: model_grid.z[0:4]
Out[9]: array([ -2.5 ,  -7.6 , -12.85, -18.25])

In [10]: model_grid.z_edges[0:4]
Out[10]: array([  0. ,  -5. , -10.2, -15.5])
```
I notice that the SOSE grid has two "edge" values of 0.0. I'm not sure what's going on here. Any ideas?

**Comment by DanJonesOcean** [comment]
Given the kind of odd structure of `sose_grid` (two zero edge values, a zero z[0] value), I'm guessing that the problem is introduced as early as `SOSEgrid`, which reads it in. Let's check these functions:
```
    # First build the model grid and check that we have the right value for split
    model_grid = grid_check_split(grid_path, split)
```
and
```
    sose_grid = SOSEGrid(sose_grid_path, model_grid=model_grid, split=split)
    # Extract land mask
    sose_mask = sose_grid.hfac == 0
```
and see where we get.

**Comment by DanJonesOcean** [comment]
The SOSE grid, as read by rdmds, looks okay:
```
RF(1) = 0
RF(2) = -10

RC(1) = -5
```
So the SOSE grid appears to get this extra zero added to it in both `RC` and `RF`.  Where is that coming from?

I'm checking within the `SOSEGrid` class. Everything looks okay when it's first read in:
```
In [11]: self.z[0:2]
Out[11]: array([ -5. , -15.5])

In [12]: self.z_edges[0:2]
Out[12]: array([  0., -10.])
```
Ooh, I bet the function `self.trim_extend` does something. In that function, these lines define the bounds:
```
            z_shallow = model_grid.z[0]
            z_deep = model_grid.z[-1]
```
producing
```
In [21]: z_shallow
Out[21]: -2.5

In [22]: z_deep
Out[22]: -6034.5500000000002
```
In this function, Depth can be extended in both directions, and the deeper bound can be trimmed. From the docs:
```
  # The indices i, j, and k will be kept track of with 4 variables each. For example, with longitude:
  # i0_before = first index we care about
  #           = how many cells to trim at beginning
  # i0_after = i0_before's position in the new grid
  #          = how many cells to extend at beginning
  # i1_before = first index we don't care about
  #           sose_nx - i1_before = how many cells to trim at end
  # i1_after = i1_before's position in the new grid
  #          = i1_before - i0_before + i0_after
  # nx = length of new grid
  #      nx - i1_after = how many cells to extend at end
```
In the next part of this function that deals with Depth, it extends the grid at the surface if `z_shallow`, which is the vertical coordinate of the center of the topmost model grid cell, is more positive than `self.z[0]`, which is the vertical coordinate of the center of the topmost SOSE grid cell:
```
# Depth
self.k0_before = 0
if z_shallow <= self.z[0]:
    # Nothing to do
    self.k0_after = 0
else:
    # Extend
    self.k0_after = 1
```
In the case of SO-WISE, the center of the topmost model grid cell is shallower than the center of the topmost SOSE grid cell, so the vertical grid is extended. That's where the extra zeros come from. The problem isn't the SOSE grid itself, it's how this routine interacts with the SOSE grid in the context of the SO-WISE grid. If I used the same topmost grid cell as SOSE, this problem would likely vanish. Here is the code that modifies the depth:
```
# Depth: can extend on both sides (depth 0 at top and extrapolated at bottom to clear the deepest model depth), trim on deep side
z_above = 0*np.ones([self.k0_after])  # Will either be [0] or empty
z_middle = self.z[self.k0_before:self.k1_before]
z_edges_middle = self.z_edges[self.k0_before:self.k1_before+1]
z_below = (2*model_grid.z[-1] - model_grid.z[-2])*np.ones([self.nz-self.k1_after])   # Will either be [something deeper than z_deep] or empty
self.z = np.concatenate((z_above, z_middle, z_below))
self.z_edges = np.concatenate((z_above, z_edges_middle, z_below))
```
But what is the purpose of this extra zero on top of the SOSE grid? It sets one of the cell centre coordinates (and an additional edge) to zero, apparently. 

This problem arose because of the small thickness of the uppermost layers in SO-WISE. If we extend them to match SOSE, perhaps with a more gradual increase, this issue should be resolved, at least in this case. However, while I can resolve this problem for SO-WISE, it does still appear to be an issue for `mitgcm_python`. It could be good to raise an issue there.

**Comment by DanJonesOcean** [comment]
I've temporarily resolved this by changing to a vertical grid with a top cell of 10 m, which matches the SOSE value. Eventually, I would like to be able to use thinner top cells, but this is an okay solution for now.

**Comment by DanJonesOcean** [comment]
Checking ICs:

- Initial temperature looks good (surface, subsurface)
- Initial salinity looks good (surface, subsurface)
- Sea ice area looks kind of weird under the shelves. Shouldn't it be zero?
- Sea ice thickness has the same kind of banded/striped structure. Looks weird. 

![seaice](https://user-images.githubusercontent.com/11757453/121547421-05262d00-ca04-11eb-8bad-fa551ad293ea.png)

**Comment by knaughten** [comment]
Regarding the extra zero in the SOSEGrid - what if this was instead a slightly positive value? A bit unphysical, but would that fix the interpolation problems for a general grid?

The stripey sea ice fields in cavities look like they're from nearest-neighbour extrapolation. Cavities should be masked to zero somewhere in the code (I think in process_ini_field), and this is the case in the SOSE initial conditions files I've made before. So there is something about the SO-WISE grid that's skipping this masking step for 2D fields. I don't think it should be a problem, because the MITgcm sea ice model should ignore anything in cavities, but might be worth looking into in case it's symptomatic of a larger problem.

**Comment by DanJonesOcean** [comment]
Thanks @knaughten. I'll give this a try, and I'll look into the sea ice masking.

**Comment by DanJonesOcean** [comment]
I'm recording this information for future reference. In my debugging and check of the sea ice initial condition routines:
- `sose_data` looks good (zero at ice shelves)
- `sose_mask` looks reasonable (zero at ice shelves)
- The field `fill` looks strange. (how is it supposed to look? maybe it's fine?)
- Both the `sose_grid` and the `model_grid` look fine
- Both `dim` and `fields` look fine
- So `model_cavity` is false everywhere. It has the dimensions of the SOSE grid though

Once inside the function `process_ini_field`, we see that `missing_cavities=True`, which is the default option. After this line:
```
source_data = discard_and_fill(source_data, source_mask[0,:], fill[0,:], use_3d=False, missing_val=missing_val)
```
The field `source_data` now looks wrong. But when I run the 'hFacC' masking step, everyting looks good again. The problem is that the ice mask is all zeros in `model_grid`. That's at least part of why it's not masking out those values. 
- `imask` and `omask` in the NetCDF file look fine...
- The code doesn't even seem to try to load an ice mask. It just gets the grid files. That's why the ice mask is zero.

I've decided to fix this manually for now, with an eye towards debugging further in the future. 

Closing for now. This issue will come up again if we try to use a top grid cell thinner than 5 m.
